ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/DataSourceExtractor.java

```diff
@@ -58,64 +53,37 @@
 
   public void extractDataSources(final AstNode node) {
     new Visitor().process(node, null);
-    commonColumnNames.addAll(Sets.intersection(leftColumnNames, rightColumnNames));
-  }
-
-  public SourceName getFromAlias() {
-    return fromAlias;
-  }
-
-  public SourceName getLeftAlias() {
-    return leftAlias;
-  }
-
-  public SourceName getRightAlias() {
-    return rightAlias;
   }
 
-  public Set<DataSource> getAllSources() {
-    return Collections.unmodifiableSet(allSources);
+  public Set<AliasedDataSource> getAllSources() {
+    return ImmutableSet.copyOf(allSources);
   }
 
   public Set<ColumnName> getCommonColumnNames() {
     return Collections.unmodifiableSet(commonColumnNames);
   }
 
-  public SourceName getFromName() {
-    return fromName;
-  }
-
-  public SourceName getLeftName() {
-    return leftName;
-  }
-
-  public SourceName getRightName() {
-    return rightName;
-  }
-
   public boolean isJoin() {
     return isJoin;
   }
 
   public SourceName getAliasFor(final ColumnName columnName) {
-    if (isJoin) {
-      if (commonColumnNames.contains(columnName)) {
-        throw new KsqlException("Column '" + columnName.text() + "' is ambiguous.");
-      }
-
-      if (leftColumnNames.contains(columnName)) {
-        return leftAlias;
-      }
-
-      if (rightColumnNames.contains(columnName)) {
-        return rightAlias;
-      }
+    if (!isJoin) {
+      return Iterables.getOnlyElement(allSources).getAlias();
+    }
 
-      throw new KsqlException(
-          "Column '" + columnName.text() + "' cannot be resolved."
-      );
+    final List<SourceName> source = allSources.stream()
+        .filter(aliased -> aliased.getDataSource().getSchema().findColumn(columnName).isPresent())
+        .map(AliasedDataSource::getAlias)
+        .collect(Collectors.toList());
+
+    if (source.size() > 1) {
```

Comment:
I feel like I've seen a lot of the same bits of code error checking column references, such as in `Analyze` where it does column name validation.  Wondering if this can be consolidated a bit.

Reply:
totally agree - I think this whole class should be removed, but that's a problem for another time :)